### PR TITLE
Add metric for disk info & last update

### PR DESCRIPTION
**Pull request by dannykopping**
This PR adds a `unused_disks_last_used_at` metric which exports the disk, PVC, and PV identifiers. The value of the gauge is the last time the disk was mounted (although this is only available on GCP for now).

I want to use this metric to create an alert for a PVC which exists without its underlying disk/volume, which happened to my team recently when we ran a script to cleanup unused disks (but it didn't clean up the associated PVCs).

To my knowledge, neither `opencost` nor `kube-state-metrics` export a metric which allows 3-way correlation between PVCs, PVs, and their underlying disks.

### CODE REVIEW

**Review by fedordikarev — APPROVED** [review]
That looks great to me 👍 
Thanks for introducing methods for Meta, that's very helpful!

**Comment by dannykopping** [comment]
Thanks @fedordikarev 
Do we need anyone else's approval before we merge?

**Review by dannykopping — COMMENTED** [review]
Heads up @Pokom @fedordikarev @inkel

**Comment by fedordikarev** [comment]
@dannykopping one review was enough for the past PRs
from what I see on changes:
- allign aws provider with other providers by adding logging
- helpers functions for Meta
- adding new prometheus metric similar to existing ones
my guess is that such changes shouldn't require much efforts (well, maybe an issue created but with detailed PR description that also could be not required here).

we may wait for @inkel review as he is initial author of the tool and introduced structure for code and modules.

File: cmd/unused-exporter/exporter.go
Comment (Pokom) [inline cmd/unused-exporter/exporter.go:159]: This is a really minor nit, I think this can be better expressed as:

```suggestion
ts := 0.0
lastUsed := d.LastUsedAt()
if !lastUsed.IsZero() {
    ts = float64(lastUsed.UnixMilli())
}				
```

File: meta.go
Comment (Pokom) [inline meta.go:97]: This is great and really like how it cleaned up the outside code. Do you think there's value in creating a small test suite for these?

**Review by Pokom — APPROVED** [review]
Left a minor nit and suggestion to add some tests for the new meta methods. Tested locally and was able to see the metrics, thanks Danny!

File: aws/provider.go
Comment (inkel) [reply aws/provider.go:56]: Nice catch! I confess I don't understand why this particular filter removes the duplicates, but I trust you 😸

File: cmd/unused-exporter/exporter.go
Comment (inkel) [reply cmd/unused-exporter/exporter.go:159]: Yup, looks much nicer, although instead of `ts := 0.0` I think the original `var ts float64` is more readable.

File: meta.go
Comment (inkel) [inline meta.go:44]: Why pulling and sorting the keys? They don't seem to be used at all. The same could be achieved in a more performant way like this:

```suggestion
	if len(m) != len(b) {
		return false
	}
```

File: meta.go
Comment (inkel) [inline meta.go:54]: How about this:
```suggestion
		if !ok || av != bv {
			return false
		}
```

File: meta.go
Comment (inkel) [reply meta.go:97]: Yes, please.

**Review by inkel — COMMENTED** [review]
Thank you @dannykopping for your contribution! In general LGTM, but I've some questions before merging this one. Also I'd like to see @Pokom comments addressed too, particularly regarding the tests for the new `Meta` methods.

File: aws/provider.go
Comment (dannykopping) [reply aws/provider.go:56]: Snapshots are returned by default when you request volumes 😒 weird behaviour by the API.

File: meta.go
Comment (dannykopping) [reply meta.go:44]: My bad, I missed something here. I should be iterating over `akeys` to see if `bkeys` exist in the reference map and its value is equivalent. I'll fix this up.

File: meta.go
Comment (dannykopping) [reply meta.go:97]: To be clear: do you just want a test suite for the `coalesce` function?
I don't see much value in testing the getters.

File: aws/provider.go
Comment (inkel) [reply aws/provider.go:56]: Ah, I see. Weird indeed!

File: meta.go
Comment (inkel) [reply meta.go:97]: `coalesce` and `Meta.Equal` 🙏🏽

File: meta.go
Comment (inkel) [reply meta.go:44]: Well, there's no need to sort the keys for that either. Just ranging on `m` will test that `b` indeed has the same keys and value.

Comment (dannykopping) [reply meta.go:44]: True, I wonder what I was thinking here 🙃 I'll try remind myself when I come back to this.

File: cmd/unused-exporter/exporter.go
Comment (Pokom) [inline cmd/unused-exporter/exporter.go:157]: @dannykopping looks like the suggestion wasn't formatted properly 😭

Comment (inkel) [reply cmd/unused-exporter/exporter.go:157]: I has spaces at the end! Fire up your Emacs, `M-x delete-trailing-whitespace` et voilà! Or just add a `before-save-hook` that runs it automatically 😬

**Comment by dannykopping** [comment]
@inkel @Pokom please have another look; I've addressed your comments

File: cmd/unused-exporter/exporter.go
Comment (inkel) [inline cmd/unused-exporter/exporter.go:169]: Any reason why this is done in a different loop than the one on line 133?

**Review by inkel — APPROVED** [review]
LGTM. Left a small question but I don't think it's blocking this.

File: cmd/unused-exporter/exporter.go
Comment (dannykopping) [reply cmd/unused-exporter/exporter.go:169]: Separation of concerns

Comment (inkel) [reply cmd/unused-exporter/exporter.go:169]: 🤔 ok. I feel like it adds more repetition and obscures the code a little bit, but as said, I don't think it's blocking at the moment.

Comment (dannykopping) [reply cmd/unused-exporter/exporter.go:169]: I disagree; I think this enhances readability by keeping each loop for a single purpose; there's nothing wrong with a little repetition.

It's a stylistic choice and you're the repo owner so if you want it changed I will change it.

Comment (inkel) [reply cmd/unused-exporter/exporter.go:169]: Nah, as said, it isn't a big of a deal. I'll leave it to @Pokom for a last review, but I'm happy to merge this.

**Review by Pokom — APPROVED** [review]

File: cmd/unused-exporter/exporter.go
Comment (Pokom) [reply cmd/unused-exporter/exporter.go:169]: It looks good to me! I don't have a strong opinion one way or another on it, and the additional tests helped me better understand the coalesce method 🙇🏻

Comment (dannykopping) [reply cmd/unused-exporter/exporter.go:169]: Thanks for the review, fellas 👍 
Can we merge this?

**Comment by inkel** [comment]
Thank you for your contribution, @dannykopping!

**Pull request merged**
